ast/ast.hpp

```diff
@@ -173,18 +179,45 @@ struct IfStatement : Statement {
     sPtr<Body> elseBody;
     bool operator==(const IfStatement& other) const {
         return Node::operator==(other) && condition == other.condition &&
-               ifBody == other.elseBody && ifBody == other.elseBody;
+               ifBody == other.ifBody && elseBody == other.elseBody;
     }
     void accept(Visitor& v) override { v.visit(this); }
 };
+
 struct Expression : Node {
+    sPtr<Expression> operand1;
+    sPtr<Expression> operand2;
+    lexer::Token operation;
+
+    bool operator==(const Expression& other) const {
+        return Node::operator==(other) && operand1 == other.operand1 &&
+               operand2 == other.operand2 && operation == other.operation;
+    }
     void accept(Visitor& v) override { v.visit(this); }
 };
-struct UnaryExpression : Expression {
+struct Primitive : Expression {
+    lexer::Token value;
+    bool operator==(const Primitive& other) const {
+        return Node::operator==(other) && value == other.value;
+    }
     void accept(Visitor& v) override { v.visit(this); }
 };
-struct BinaryExpression : Expression {
+struct ModifiablePrimary : Expression {
+    std::vector<sPtr<Expression>> args;
```

Comment:
What kind of values is `args` supposed to hold?

Reply:
Everything, which is called after the `.`, or inside `[]`, mixed up with the operations. I agree, that the structure is messy.

Reply:
If you can suggest better structure - tell me, which one

Reply:
We can consider member access (whether `.` or `[]`) to be a binary operation. If we make the dot-access even a more specific operation, then the second operand would be limited to just an identifier and not any expression (an `Identifier` can sub-class `Expression`).

Reply:
@MefAldemisov you can think that `.` and `[` are binary operators. In an if-statement, if it's the first one, you'll expect an identifier, and in the latter case, you'll expect an expression and a closing bracket.